scripts/generate_upgrade_syncset.py

```diff
@@ -0,0 +1,301 @@
+
+            except OSError as err:
+                logging.info('Command \'{}\' failed with error: {}'.format(CLUSTER_LIST_EXEC, str(err)))
+                sys.exit(1)
+            except subprocess.CalledProcessError as err:
+                logging.info('Command \'{}\' failed with error: {}'.format(CLUSTER_LIST_EXEC, str(err)))
+                sys.exit(1)
+            except ValueError as err:
+                logging.info(
+                    'Output of command \'{}\' failed to parse with error: {}'.format(CLUSTER_LIST_EXEC, str(err)))
+                sys.exit(1)
+
+    return cluster_id_to_name
+
+
+def get_cluster_name_for_id(cluster_id, environment):
+    """
+    Determines the clusterdeployment name associated with the supplied clusterid
+    :param cluster_id: cluster-id of cluster
+    :return: clusterdeployment name associated with cluster ID
+    """
+    cluster_ns = 'uhc-{}-{}'.format(environment, cluster_id)
+    try:
+        # pout = subprocess.check_output(['cat', '/home/mbargenq/scratch/cluster.txt']).decode(
+        #     'utf-8')
+        pout = subprocess.check_output(['oc', 'get', 'clusterdeployment', '-n', cluster_ns, '-o', 'json']).decode(
+            'utf-8')
+        output_as_dict = json.loads(pout)
+        if 'items' not in output_as_dict:
+            raise Exception('fetching clusterdeployment returned invalid response')
+
+        cd = output_as_dict['items']
+        if len(cd) == 0:
+            raise Exception('clusterdeployment not found')
+
+        return cd[0]['metadata']['name']
+
+    except Exception as err:
+        raise err
+
+
+def generate_upgradeconfig(start_time, version, channel):
+    """
+    Generates the UpgradeConfig CR in python dict form
+    :param version: version to upgrade to
+    :param channel: channel for upgrade
+    :param force: force upgrade
+    :return: dict representation of upgradeconfig CR
+    """
+    uc = {
+        'apiVersion': UC_API_VERSION,
+        'kind': 'UpgradeConfig',
+        'metadata': {
+            'name': UC_NAME,
+            'namespace': UC_NAMESPACE,
+        },
+        'spec': {
+            'type': dq('OSD'),
+            'upgradeAt': dq(start_time),
+            'proceed': True,
+            'PDBForceDrainTimeout': UC_PDB_DRAIN_TIMEOUT,
+            'desired': {
+                'version': dq(version),
+                'channel': dq(channel),
+                'force': UC_FORCE,
+            }
+        }
+    }
+    return uc
+
+
+def generate_syncset(upgradeconfig, cluster_id, cluster_name='', environment='production'):
+    """
+    Generates the SelectorSyncSet
+    :param upgradeconfig: UpgradeConfig CR as dict
+    :param cluster_id: Cluster ID to target
+    :param cluster_name: Cluster name to target
+    :return: SelectorSyncSet as dict
+    """
+
+    # generate the namespace for the SyncSet in hive format
+    # uhc-<environment>-<clusterID>
+    ss_namespace = 'uhc-{}-{}'.format(environment, cluster_id)
+
+    data = CM()
```

Comment:
Is there a native Python `SyncSet` object that can be used instead of crafting a YAML document? The hope is that the native object will wrap the construction of the object and provide guardrails ensuring that required attributes are present and of the correct type.

Reply:
I had a look around but was unable to find any helper packages unfortunately. I'm hoping to eventually be able to use the cluster_mgmt API instead of this clunky via-Hive method (which would eliminate the need to be hand-rolling a SyncSet), but that's still TBD pending discussions with the SDA. Doing it this way will be the contingency.